### ISSUE TITLE
Multi-gpu training

### ISSUE THREAD

**Issue by JohnGiorgi** [KevinMusgrave/pytorch-metric-learning#10]
Hi, I am stuck on how multi-GPU training would work for loss functions with more than one negative, particularly [NTXentLoss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#ntxentloss). 

In [SimCLR](https://arxiv.org/abs/2002.05709), the number of _negatives_ per some _positive_ pair is taken to be `2 * (N - 1)` (all examples in a minibatch of size `N` that don't belong to that positive pair), and they find ([as other works before them](https://github.com/KevinMusgrave/pytorch-metric-learning)) that the bigger the batch size, the larger the number of negatives, and the better the learned representations.

`DataParallel` and `DistributedDataParallel` divide up a mini-batch, send each partition of examples to a GPU, compute gradients, and then average these gradients before backpropagating. But this means that each GPU is computing a loss with `N/n_gpus` examples and, therefore,  `2 * (N/n_gpus - 1)` negatives per positive pair.

My question is: how might I use a loss function from this library, such as [NTXentLoss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#ntxentloss), with `DataParallel` or `DistributedDataParallel` that avoids this "issue"? I.e. that allows me to use multiple GPUs while maintaining `2 * (N - 1)` negatives per positive pair.

**Comment by KevinMusgrave** [comment]
DataParallel should work fine, because it's still just 1 process running. This is what I always use.

I haven't used DistributedDataParallel. I think you're right that each process will only see part of the batch, and therefore the number of positive and negative pairs will be wrong. At the moment I can't think of an easy way to fix that. Maybe the loss function in one process needs to get embeddings from all the other processes?

**Comment by JohnGiorgi** [comment]
Hmm, okay. I will have to look into `DataParallel` more carefully. The [PyTorch documentation](https://pytorch.org/docs/1.3.0/nn.html#torch.nn.DataParallel) suggests it is still splitting up batches across devices, so I would suspect that the problem I mentioned above still stands:

> This container parallelizes the application of the given module by splitting the input across the specified devices by chunking in the batch dimension (other objects will be copied once per device). In the forward pass, the module is replicated on each device, and each replica handles a portion of the input. During the backwards pass, gradients from each replica are summed into the original module

Is [CrossBatchMemory](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#crossbatchmemory) a potential solution here?

**Comment by KevinMusgrave** [comment]
Yes, if you wrap your model with DataParallel, the when computing the output, the batch will be split across devices. But if you do not wrap the loss function with DataParallel, then the loss function will have access to the entire batch. So it will be able to compute all negative pairs. 

In contrast, I think DistributedDataParallel will launch multiple separate processes. So the loss function in each process will not have access to the entire batch by default.

**Comment by KevinMusgrave** [comment]
CrossBatchMemory stores embeddings from previous iterations and creates pairs using the current iteration's embeddings combined with those previous embeddings. So it works with the assumption that embeddings drift slowly as training progresses. You can try it, but it will be very different from the SimCLR paper. Also, it's really slow right now and I need to fix that.

**Comment by KevinMusgrave** [comment]
Here's something you can check. 
```python
from pytorch_metric_learning import losses, miners
# set margins to ensure that no pairs are left out for this example
mining_func = miners.PairMarginMiner(pos_margin=0, neg_margin=100, use_similarity=False)

# in your training loop:
pairs = mining_func(embeddings, labels)
print(mining_func.num_pos_pairs)
print(mining_func.num_neg_pairs)
```
This will print the number of positive and negative pairs "mined", but really it's just the total number of positive and negative pairs because the pos_margin and neg_margin are set to extreme values. As long as you don't wrap the mining function with DataParallel then it should print the correct number. You can still wrap your model with DataParallel though.

**Comment by JohnGiorgi** [comment]
Trying to wrap my head around this example.

I generated a dummy batch of 32 positive pairs, each of dim 128. The printout gives me the expected number of positive pairs (32) but an unexpected number of negative pairs (960). I would have expected `2(batch_size - 1) = 2(32 - 1) = 62`.

Any idea what I am misunderstanding?

```python
import torch
from pytorch_metric_learning import losses, miners

batch_size = 16
embedding_dim = 128

# generate a dummy batch
anchor_embeddings = torch.randn(batch_size, embedding_dim)
positive_embeddings = torch.randn(batch_size, embedding_dim)
embeddings = torch.cat((anchor_embeddings, positive_embeddings),)
indices = torch.arange(0, anchor_embeddings.size(0), device=anchor_embeddings.device)
labels = torch.cat((indices, indices))

# set margins to ensure that no pairs are left out for this example
mining_func = miners.PairMarginMiner(pos_margin=0, neg_margin=100, use_similarity=False)

# in your training loop:
pairs = mining_func(embeddings, labels)
print(mining_func.num_pos_pairs)  #=> 32
print(mining_func.num_neg_pairs)  #=> 960
```

**Comment by KevinMusgrave** [comment]
@JohnGiorgi There are 32 samples. For each sample, 1/31 are positives, and 30/31 are negatives. So 30*32 = 960

**Comment by JohnGiorgi** [comment]
Right. Thank you!

**Comment by JohnGiorgi** [comment]
Okay, I think I have figured out how this works with `DistributedDataParallel`. Writing it up here in case anyone else stumbles on this problem, as it is non-trivial to solve:

You need to:

1. `gather_all` `embeddings` from all replicas.
2. Because gathered tensors have no gradients, we overwrite the gathered `embeddings` tensor from the current replica with the `embeddings` tensor produced on that replica, which has gradients.
3. Concatenate the list of `embeddings` before computing the loss.

Some almost functional pseudo-code:

```python
import torch
import torch.distributed as dist

# Gather the embeddings from every replica. These would have been created by feeding batches of text through some sort of encoder.
embeddings_list = [torch.ones_like(embeddings) for _ in range(dist.get_world_size())]
dist.all_gather(embeddings_list, embeddings)

# Gathered tensors have no gradient, so we set the position in the list equal to the current replicas embeddings, which do have gradients
embeddings_list[dist.get_rank()] = embeddings

# Finally, concatenate the list of embeddings before computting a loss
embeddings = torch.cat(embeddings_list)

loss = some_pytorch_metric_learning_loss(embeddings, labels)
```

In my application, this works. And the trick you mention to print out the `num_neg_pairs` from above prints the expected value.

**Comment by cbaziotis** [comment]
@JohnGiorgi Thanks for sharing your solution. I was wondering how your approach differs from other approaches that subclass [`torch.autograd.Function`](https://pytorch.org/docs/stable/autograd.html#function), such as:

1.  https://github.com/Spijkervet/SimCLR/blob/master/simclr/modules/nt_xent.py that uses this [Function](https://github.com/Spijkervet/SimCLR/blob/master/simclr/modules/gather.py).
2. https://github.com/PyTorchLightning/lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py#L224 that uses this [Function](https://github.com/PyTorchLightning/lightning-bolts/blob/5577453a6d7072724d9ae24184daf8f45d4baff7/pl_bolts/models/self_supervised/simclr/simclr_module.py#L20).

Is there a practical difference or is it just a matter of style?

Thanks!

**Comment by weiyx16** [comment]
> @JohnGiorgi Thanks for sharing your solution. I was wondering how your approach differs from other approaches that subclass [`torch.autograd.Function`](https://pytorch.org/docs/stable/autograd.html#function), such as:
> 
> 1. https://github.com/Spijkervet/SimCLR/blob/master/simclr/modules/nt_xent.py that uses this [Function](https://github.com/Spijkervet/SimCLR/blob/master/simclr/modules/gather.py).
> 2. https://github.com/PyTorchLightning/lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py#L224 that uses this [Function](https://github.com/PyTorchLightning/lightning-bolts/blob/5577453a6d7072724d9ae24184daf8f45d4baff7/pl_bolts/models/self_supervised/simclr/simclr_module.py#L20).
> 
> Is there a practical difference or is it just a matter of style?
> 
> Thanks!
> 
> **Update**: This [blog post](https://amsword.medium.com/gradient-backpropagation-with-torch-distributed-all-gather-9f3941a381f8) claims that using your approach, we should re-scale the gradients by the number of `word_size` to obtain the correct gradients.

I have taken an experiment on a small toy example and compared the grad between two methods, and I think these two ways create same gradients.

**Comment by chenxi52** [comment]
Do you think  this version --- https://github.com/Spijkervet/SimCLR/blob/master/simclr/modules/nt_xent.py  should   re-scale the gradients ???  if its parameters works in a single gpu too( i think so
